narwhal/worker/src/batch_fetcher.rs

```diff
@@ -0,0 +1,530 @@
+                    self.metrics
+                        .worker_batch_fetch
+                        .with_label_values(&["remote", "success"])
+                        .inc();
+                    debug!("Found {} batches remotely", remote_batches.len());
+                    for (remote_batch_digest, remote_batch) in remote_batches {
+                        if digests.contains(&remote_batch_digest) {
+                            fetched_batches.insert(remote_batch_digest, remote_batch);
+                        }
+                    }
+                    return fetched_batches;
+                }
+                Err(err) => {
+                    if err.to_string().contains("Timeout") {
+                        self.metrics
+                            .worker_batch_fetch
+                            .with_label_values(&["remote", "timeout"])
+                            .inc();
+                    } else {
+                        self.metrics
+                            .worker_batch_fetch
+                            .with_label_values(&["remote", "fail"])
+                            .inc();
+                    }
+                    debug!("Error retrieving payloads {digests:?} from {worker} attempt {attempt}: {err}")
+                }
+            }
+            tokio::time::sleep(timeout).await;
+        }
+    }
+
+    /// Issue request_batches RPC and verifies response integrity
+    async fn safe_request_batches(
+        &self,
+        digests: HashSet<BatchDigest>,
+        worker: NetworkPublicKey,
+        timeout: Duration,
+    ) -> anyhow::Result<HashMap<BatchDigest, Batch>> {
+        let mut verified_batches = HashMap::new();
+        let mut digests_to_fetch = digests.clone();
+
+        // Continue to bulk request from remote worker until no remaining
+        // digests are available.
+        loop {
+            if digests_to_fetch.is_empty() {
+                break;
+            }
+            let RequestBatchesResponse {
+                batches,
+                is_size_limit_reached,
+            } = self
+                .network
+                .request_batches(
+                    digests_to_fetch.clone().into_iter().collect(),
+                    worker.clone(),
+                    timeout,
+                )
+                .await?;
+            // Use this flag to determine if the worker returned a valid digest
+            // to protect against an adversarial worker returning size limit reached
+            // and no batches leading us to retry this worker forever or until
+            // we get batches from other workers.
+            let mut is_digest_received = false;
+            for batch in batches {
+                let batch_digest = batch.digest();
+                if !digests_to_fetch.contains(&batch_digest) {
+                    bail!("[Protocol violation] Worker {worker} returned batch with digest {batch_digest} which is not part of the requested digests: {digests:?}");
```

Comment:
Maybe we dont want to bail here? A byzantine worker could return a incorrect batch on purpose